### PR TITLE
Geochemistry functionality enhanced

### LINKED ISSUE #14948: Create MinimalGeochemistryDatabase
## Reason

The GeochemicalDatabaseReader reads everything in the database.   But geochemistry models always use a subset of this data.  Create a structure or class that holds this "actually-used" data in suitable data structures that can be rapidly accessed and manipulated.

## Design

Two important aspects are:

- the MinimalGeochemistryDatabase's constructor should perform suitable checks that the user has requested sensible information from the entire database.  For example, one of the basis species must always be H2O.  I have a long list of requirements, but will not write them all here.  All the equilibrium species should be extracted from the DatabaseReader automatically.

- the data structures used in MinimalGeochemistryDatabase must be rapidly accessible.  Almost definitely, that means DenseMatrix and std::vector, rather than the std::maps that DatabaseReader uses.  These data structures will be queried many times per node per timestep.

## Impact

MinimalGeochemistryDatabase will act as the main interface between "normal" MOOSE code (Kernels, etc) and the database.

I've actually already coded this and am just waiting for #14911 to be merged.  Tagging @cpgr.

### LINKED ISSUE #14976: free electron in Geochemistry
## Reason

Include the free electron in the database converters and Reader because it is needed to compute pe and Eh

## Design

free electron added to python convert utilities. GeochemicalDatabaseReader stores this information as a secondary species. moose_testdb.json has free electron in it to enable testing.

Unfortunately, the free electron is written in terms of O2(g) in the GWB database, and O2(g) is not a basis species.  This is probably not a typo, but instead something that's suboptimal.  Perhaps programs like the Geochemists Workbench automatically add O2(g) and e- if O2(aq) is a basis species.  There are two options

1. manually modify the database to include O2(aq) instead of O2(g) (changing the equilibrium constants).  Oh dear, will we always remember to do this?

2. include some "if basis is O2(aq) then include e- as an equilibrium species, writing its equilibrium constants in terms of O2(aq) instead of O2(g), and maybe include O2(g) as a gas species too".  Oh dear, this is a "secret" thing going on behind the scenes that not all users will be aware of.  It does (1) automatically.

The example of Section 11.3 of the Bethke book works numerically if (1) is implemented (and would do if we did (2) as well) which demonstrates that the O2(g) is not a typo.

## Impact

Ability to compute pe and Eh.  Tagging @cpgr .

### LINKED ISSUE #14979: add MOOSE version of GWB geochem database
## Reason

I think this will be very useful for `geochemistry` users as they'll be able to use a centrally-maintained database in their simulations.

## Design

Super easy - just use database_converter on the GWB database

## Impact

Lots of extra lines in the MOOSE repo!

### LINKED ISSUE #14981: Geochemistry: add MOOSE action to query database
## Reason

Some simple "simulations" actually just involve querying and manipulating the database to find equilibrium constants, equilibrium temperatures, reactions, etc.  This should be possible to do in MOOSE.

## Design

Make UserObjects that parse the database, make a minimal system, do species swaps.  Make an Action that adds required userobjects and does the "simulation" (there is no time and space, hence the quotes around "simulation" - it's all very simple).

## Impact

Ability to extract useful information quickly and manipulate the database using MOOSE.

### PR DESCRIPTION
**Pull request by WilkAndy**
This is just #14978 all squashed into one commit

- free electron GWB format now parsed by gwb_reader
- free electron included in dbclass and dbconverter
- free electron and O2(g) added to unit/data/moose_testdb.json
- free electron added to GeochemicalDatabaseReader as just another secondary species
- GeochemicalDatabaseReader has the option that during database parsing the free electron can be rexpressed in terms of O2(aq) instead of O2(g).
- all python and C++ tested
Fixes #14976

data/moose_geochemdb.json added - lots of lines, wow.
Fixes #14979

MinimalGeochemicalSystem and unit tests added
This builds the minimal geochemical system of basis species, equilibium species and kinetic species that is relevant for use-supplied basis components, mineral, gases, etc.  The data is stored in a computationally-efficient way using std::vector and DenseMatrix.
Also, add to moose_testdb.json to allow testing for wacky cases.
Fixes #14948

GeochemistrySpeciesSwapper utility added
ixes #14949

GeochemicalModelRoot Userobject added: this provides a MOOSE interface to utils/MinimalGeochemicalSystem

GeochemicalModelInterrogator Userobject added: this enables querying and simple manipulations of the database
GeochemicalModelInterrogator Action added: this adds the GeochemicalModelInterrogator Userobject as well as setting up a simple Moose simulation with no timestepping, no output, etc

Lots of documentation and examples illustrating this functionality

Tests of the functionality

Fixes #14981

add GeochemistryApp to base files. Refs #14984

Changes requested by @cpgr and @loganharbour all done

Refs #14984
Refs #14981

Refs #14981

### CODE REVIEW

**Review by loganharbour — CHANGES_REQUESTED** [review]
All very minor changes. Should be good after this. Looks like the output object was pretty straight forward!

File: modules/geochemistry/test/tests/interrogate_reactions/muscovite2.i [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  swap_out_of_basis = "SiO2(aq) Al+++   Quartz   SiO2(aq)  Tridymite SiO2(aq)"
+  swap_into_basis = "  Quartz   Maximum SiO2(aq) Tridymite SiO2(aq)  Amrph^silica"
+  equilibrium_species = Muscovite
+  interrogation = activity
+[]
+
+[UserObjects]
+  [./definition]
+    type = GeochemicalModelDefinition
+    database_file = "../../../database/moose_geochemdb.json"
```
[the remark this answers is not in the document]
Comment (WilkAndy) [reply modules/geochemistry/test/tests/interrogate_reactions/muscovite2.i:13]: Yea, OK.  Let's leave this for another PR, as you say.   I'm still wondering how many databases we'll have (hopefully only 1) and about their organisation

**Comment by WilkAndy** [comment]
Oh yea, all tests passed.  Good to merge, @loganharbour  ?

**Review by loganharbour — APPROVED** [review]

**Comment by WilkAndy** [comment]
Woohoo, thanks @loganharbour !

**Pull request merged**